version: "3.8"

services:

  redis:
    image: redis:latest
    container_name: redis
    restart: on-failure
    healthcheck:
      test: [ "CMD", "redis-cli", "ping" ]
      interval: 5s
      timeout: 3s
      retries: 5
    expose:
      - "6379"

  postgres:
    image: postgres:16.3-alpine
    container_name: postgres
    restart: unless-stopped
    env_file:
      - .env
    environment:
      POSTGRES_DB: ${DB_NAME}
      POSTGRES_USER: ${DB_USER}
      POSTGRES_PASSWORD: ${DB_PASSWORD}
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready -U ${DB_USER}" ]
      interval: 5s
      timeout: 5s
      retries: 5
    expose:
      - "5432"
    volumes:
      - postgres_data:/var/lib/postgresql/data

  django:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: django
    command: >
      sh -c "
      python3 manage.py migrate &&
      python3 manage.py loaddata data.json &&
      python3 manage.py runserver 0.0.0.0:8000
      "
    ports:
      - "8000:8000"
    volumes:
      - ./:/app/
    env_file:
      - .env
    environment:
      DB_HOST: postgres
    depends_on:
      postgres:
        condition: service_healthy

  celery:
    container_name: celery_worker
    build: .
    tty: true
    command: celery -A habit_reminder worker --loglevel=info
    restart: on-failure
    env_file:
      - .env
    volumes:
      - ./:/app/
    depends_on:
      - redis
      - django
      - postgres

  celery-beat:
    container_name: celery_beat
    build: .
    tty: true
    command: celery -A habit_reminder beat --loglevel=info
    restart: on-failure
    env_file:
      - .env
    volumes:
      - ./:/app/
    depends_on:
      - redis
      - django
      - postgres

volumes:
  postgres_data:
